write-fonts/src/tables/cmap.rs

```diff
@@ -4,78 +4,45 @@
 
 include!("../../generated/generated_cmap.rs");
 
+use std::collections::HashMap;
+
 // https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#windows-platform-platform-id--3
 const WINDOWS_BMP_ENCODING: u16 = 1;
+const WINDOWS_FULL_REPERTOIRE_ENCODING: u16 = 10;
 
 // https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#unicode-platform-platform-id--0
 const UNICODE_BMP_ENCODING: u16 = 3;
+const UNICODE_FULL_REPERTOIRE_ENCODING: u16 = 4;
 
 fn size_of_cmap4(seg_count: u16, gid_count: u16) -> u16 {
     8 * 2  // 8 uint16's
     + 2 * seg_count * 4  // 4 parallel arrays of len seg_count, 2 bytes per entry
     + 2 * gid_count // 2 bytes per gid in glyphIdArray
 }
 
-impl CmapSubtable {
-    fn create_format_4(
-        lang: u16,
-        end_code: Vec<u16>,
-        start_code: Vec<u16>,
-        id_deltas: Vec<i16>,
-    ) -> Self {
-        assert!(
-            end_code.len() == start_code.len() && end_code.len() == id_deltas.len(),
-            "uneven parallel arrays, very bad. Very very bad."
-        );
-
-        let seg_count: u16 = start_code.len().try_into().unwrap();
-        // Spec: Log2 of the maximum power of 2 less than or equal to segCount (log2(searchRange/2),
-        // which is equal to floor(log2(segCount)))
-        let entry_selector = (seg_count as f32).log2().floor();
-
-        // Spec: Maximum power of 2 less than or equal to segCount, times 2
-        // ((2**floor(log2(segCount))) * 2, where “**” is an exponentiation operator)
-        let search_range = 2u16.pow(entry_selector as u32).checked_mul(2).unwrap();
-
-        // if 2^entry_selector*2 is a u16 then so is entry_selector
-        let entry_selector = entry_selector as u16;
-        let range_shift = seg_count * 2 - search_range;
-
-        let id_range_offsets = vec![0; id_deltas.len()];
-        CmapSubtable::Format4(Cmap4::new(
-            size_of_cmap4(seg_count, 0),
-            lang,
-            seg_count * 2,
-            search_range,
-            entry_selector,
-            range_shift,
-            end_code,
-            start_code,
-            id_deltas,
-            id_range_offsets,
-            vec![], // becauseour idRangeOffset's are 0 glyphIdArray is unused
-        ))
-    }
+fn size_of_cmap12(num_groups: u32) -> u32 {
```

Comment:
Optional, in re-reading I tend to think both size_of fns should have a link to the cmap spec

Reply:
Done